src/main/java/com/arextest/common/interceptor/MetricInterceptor.java

```diff
@@ -0,0 +1,59 @@
+package com.arextest.common.interceptor;
+
+import static com.arextest.common.utils.MetricUtils.*;
+import com.arextest.common.metrics.MetricListener;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.apache.commons.collections4.CollectionUtils;
+import org.springframework.stereotype.Component;
+import org.springframework.web.servlet.HandlerInterceptor;
+
+/**
+ * record request and response interceptor
+ * created by xinyuan_wang on 2023/12/25
+ */
+@Component
+public class MetricInterceptor implements HandlerInterceptor {
+  public final List<MetricListener> metricListeners;
+
+  public MetricInterceptor(List<MetricListener> metricListeners) {
+    this.metricListeners = metricListeners;
+  }
+
+  @Override
+  public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler) {
+    request.setAttribute(START_TIME, System.currentTimeMillis());
+    return true;
+  }
+
+  @Override
+  public void afterCompletion(HttpServletRequest request, HttpServletResponse response, Object handler, Exception ex)
+      throws IOException {
+    long endTime = System.currentTimeMillis();
+    long startTime = (Long) request.getAttribute(START_TIME);
```

Comment:
`long startTime = (long) request.getAttribute(START_TIME);`
if ensure result is not null, reduce one unbox